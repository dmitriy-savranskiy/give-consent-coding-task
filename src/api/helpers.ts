export const BASE_URL = '/api/';

export function url(url: string, base = BASE_URL): string {
  const baseWithoutEndingSlash = base.replace(/\/$/, '');
  const urlWithoutStartingSlash = url.replace(/^\//, '');

  return `${baseWithoutEndingSlash}/${urlWithoutStartingSlash}`;
}
