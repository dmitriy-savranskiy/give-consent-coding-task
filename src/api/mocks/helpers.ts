const IS_TEST_ENV = process.env.NODE_ENV === 'test';

export function saveToSessionStorage(key: string, value: unknown): void {
  const json = JSON.stringify(value);

  sessionStorage.setItem(key, json);
}

export function getFromSessionStorage<T>(key: string, fallback?: T): T | null {
  if (IS_TEST_ENV) {
    return fallback || null;
  }

  const json = sessionStorage.getItem(key);

  if (!json) {
    return null;
  }

  return JSON.parse(json);
}

export function getUniqueId(): string {
  const sessionStorageKey = 'http-mock-unique-id';
  const lastId = sessionStorage.getItem(sessionStorageKey) || '0';
  const newId = parseInt(lastId) + 1;

  sessionStorage.setItem(sessionStorageKey, String(newId));

  return String(newId);
}

export function delay(delay: number): number {
  return IS_TEST_ENV ? 0 : delay;
}
