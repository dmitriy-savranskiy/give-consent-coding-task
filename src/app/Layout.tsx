import { ListItem, List, Container } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { NavLink } from 'react-router-dom';

export type LayoutProps = {
  children: JSX.Element;
};

const links = [
  {
    link: '/give-consent',
    label: 'Give consent',
  },
  {
    link: '/consents',
    label: 'Collected consents',
  },
] as const;

const useStyles = makeStyles((theme) => ({
  navLink: {
    ...theme.typography.button,
    textDecoration: 'none',
    color: theme.palette.text.primary,
  },
  navLinkActive: {
    fontWeight: 'bold',
  },
  navContainer: {
    borderRight: `1px solid ${theme.palette.divider}`,
    height: '100vh',
  },
  layoutWrapper: {
    display: 'flex',
  },
  content: {
    padding: '1rem',
    flexGrow: 1,
  },
}));

export function Layout(props: LayoutProps): JSX.Element {
  const classes = useStyles();

  return (
    <div className={classes.layoutWrapper}>
      <nav className={classes.navContainer}>
        <List>
          {links.map(({ link, label }) => (
            <NavLink
              key={link}
              className={classes.navLink}
              activeClassName={classes.navLinkActive}
              to={link}
            >
              <ListItem button>{label}</ListItem>
            </NavLink>
          ))}
        </List>
      </nav>
      <main className={classes.content}>
        <Container>{props.children}</Container>
      </main>
    </div>
  );
}
