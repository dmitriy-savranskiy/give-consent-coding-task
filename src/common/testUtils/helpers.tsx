import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { createMemoryHistory, MemoryHistory } from 'history';
import { store } from '../../app/store';

export function renderWithStore(
  ui: JSX.Element
): ReturnType<typeof render> & { history: MemoryHistory } {
  const history = createMemoryHistory()
  const renderResult = render(
    <Router history={history}>
      <Provider store={store}>{ui}</Provider>
    </Router>
  );

  return {
    ...renderResult,
    history,
  };
}

export function mockComponentForSnapshot(name: string) {
  return (props: { children?: JSX.Element } & unknown): JSX.Element => {
    const { children, ...restProps } = props;

    return (
      <div>
        <div>{name}Mock</div>
        <div>{JSON.stringify(restProps)}</div>
        <div>{children}</div>
      </div>
    );
  };
}

// use this function only when you know that the input element is rendered no matter what.
export function getInputEl(element: HTMLElement): HTMLInputElement {
  return element.querySelector('input') as HTMLInputElement;
}
