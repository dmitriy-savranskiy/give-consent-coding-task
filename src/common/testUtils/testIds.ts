// consent targets
export const tidConsentTarget = (targetId: string) =>
  `CONSENT-TARGET-${targetId}`;
export const tidNoConsents = () => `NO-CONSENTS`;

// give consent
export const tidGiveConsentName = () => `GIVE-CONSENT-NAME`;
export const tidGiveConsentEmail = () => `GIVE-CONSENT-EMAIL`;
export const tidGiveConsentConsentTargets = () =>
  `GIVE-CONSENT-CONSENT-TARGETS`;
export const tidGiveConsentGiveBtn = () => `GIVE-CONSENT-GIVE-BTN`;

// common
export const tidLoader = () => `LOADER`;
