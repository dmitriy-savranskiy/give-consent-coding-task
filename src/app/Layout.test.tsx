import renderer from 'react-test-renderer';
import { mockComponentForSnapshot } from '../common/testUtils/helpers';
import { Layout } from './Layout';

jest.mock('@material-ui/core', () => {
  return {
    ListItem: mockComponentForSnapshot('ListItem'),
    List: mockComponentForSnapshot('List'),
    Container: mockComponentForSnapshot('Container'),
  };
});

jest.mock('react-router-dom', () => {
  return {
    NavLink: mockComponentForSnapshot('NavLink'),
  };
});

describe('Layout component', () => {
  it('should match snapshot', () => {
    const component = renderer.create(
      <Layout>
        <span>content</span>
      </Layout>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
