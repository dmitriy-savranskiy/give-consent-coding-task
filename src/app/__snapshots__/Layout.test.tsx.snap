// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component should match snapshot 1`] = `
<div
  className="makeStyles-layoutWrapper-4"
>
  <nav
    className="makeStyles-navContainer-3"
  >
    <div>
      <div>
        List
        Mock
      </div>
      <div>
        {}
      </div>
      <div>
        <div>
          <div>
            NavLink
            Mock
          </div>
          <div>
            {"className":"makeStyles-navLink-1","activeClassName":"makeStyles-navLinkActive-2","to":"/give-consent"}
          </div>
          <div>
            <div>
              <div>
                ListItem
                Mock
              </div>
              <div>
                {"button":true}
              </div>
              <div>
                Give consent
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            NavLink
            Mock
          </div>
          <div>
            {"className":"makeStyles-navLink-1","activeClassName":"makeStyles-navLinkActive-2","to":"/consents"}
          </div>
          <div>
            <div>
              <div>
                ListItem
                Mock
              </div>
              <div>
                {"button":true}
              </div>
              <div>
                Collected consents
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <main
    className="makeStyles-content-5"
  >
    <div>
      <div>
        Container
        Mock
      </div>
      <div>
        {}
      </div>
      <div>
        <span>
          content
        </span>
      </div>
    </div>
  </main>
</div>
`;
